tech/post/2017-09-25-Kubernetes-Scheduler.md

```diff
@@ -0,0 +1,145 @@
+	"hardPodAffinitySymmetricWeight" : 10
+}
+```
+
+## 4. Scheduler Algorithm
+
+### 4.1 Default Predicates
+
+#### 4.1.1 NoVolumeZoneConflict
+
+Evaluate if the volumes a pod requests are available on the node, given the Zone restrictions.
+
+#### 4.1.2 MaxEBSVolumeCount
+Ensure that the number of attached ElasticBlockStore volumes does not exceed a maximum value (by default, 39, since Amazon recommends a maximum of 40 with one of those 40 reserved for the root volume -- see Amazon's documentation). The maximum value can be controlled by setting the KUBE_MAX_PD_VOLS environment variable.
+
+#### 4.1.3 MaxGCEPDVolumeCount
+Ensure that the number of attached GCE PersistentDisk volumes does not exceed a maximum value (by default, 16, which is the maximum GCE allows -- see GCE's documentation). The maximum value can be controlled by setting the KUBE_MAX_PD_VOLS environment variable.
+
+#### 4.1.4 MaxAzureDiskVolumeCount
+Ensure that the number of attached Azure volumes does not exceed a maximum value (by default, 16).
+
+#### 4.1.5 MatchInterPodAffinity
+Fit is determined by inter-pod affinity. AffinityAnnotationKey represents the key of affinity data (json serialized) in the Annotations of a Pod.
+
+#### 4.1.6 NoDiskConflict
+Evaluate if a pod can fit due to the volumes it requests, and those that are already mounted. Currently supported volumes are: AWS EBS, GCE PD, ISCSI and Ceph RBD. Only Persistent Volume Claims for those supported types are checked. Persistent Volumes added directly to pods are not evaluated and are not constrained by this policy.
+
+#### 4.1.7 GeneralPredicates
+* PodFitsResources: 
+Check if the free resource (CPU and Memory) meets the requirement of the Pod. The free resource is measured by the capacity minus the sum of requests of all Pods on the node. To learn more about the resource QoS in Kubernetes, please check QoS proposal.
+
+* PodFitsHost: 
+Filter out all nodes except the one specified in the PodSpec's NodeName field.
+
+* PodFitsHostPorts: 
+Check if any HostPort required by the Pod is already occupied on the node.
+
+* PodMatchNodeSelector: 
+PodMatchNodeSelector checks if a pod node selector matches the node label.
+
+#### 4.1.8 CheckNodeMemoryPressure
+Check if a pod can be scheduled on a node reporting memory pressure condition. Currently, no BestEffort should be placed on a node under memory pressure as it gets automatically evicted by kubelet.
+
+#### 4.1.9 CheckNodeDiskPressure
+Check if a pod can be scheduled on a node reporting disk pressure condition. Currently, no pods should be placed on a node under disk pressure as it gets automatically evicted by kubelet.
+
+#### 4.1.10 NoVolumeNodeConflict
+Fit is determined by volume zone requirements.
+
+#### 4.1.11 PodToleratesNodeTaints
+Fit is determined based on whether a pod can tolerate all of the node's taints.
+
+### 4.2 Default Priorities
+```FinalScoreNode = (Weight1 * priorityFunc1) + (Weight2 * priorityFunc2) + ... + (Weightn * priorityFuncn)```

```

Comment:
``priorityFunc1`` stands for the result of this function?

Reply:
Yeah. priorityFunc n represent all the check functions.